### DISCUSSION TITLE
Should node imply ESM when importing via Export Map aka Subpath Exports?

### DISCUSSION THREAD

**Discussion by mattfysh** [nodejs/help#4856] [Q&A]
See reproduction: https://github.com/mattfysh/node-subpath-exports

I would assume that when the top-level package does `import 'pkg'`, and the import specifier is resolved via the export map in `node_modules/pkg/package.json` under the "import" condition, that the imported module should be assumed to be ESM

I believe it should be assumed to be ESM because of the condition under which it was imported.

Instead I get this error:

```
/node-subpath-exports/node_modules/pkg/main.esm.js:1
import util from 'util'
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

**Comment by aduh95** [comment]
Ping @nodejs/modules 

> I believe it should be assumed to be ESM because of the condition under which it was imported.

`"import"` could resolve to a JSON or WASM module, it would be wrong to assume it's always ECMAScript IMHO.

**Reply by ljharb** [reply]
you can also import CJS.

The module author, NEVER the importer, is the only person who can authoritatively determine the format of a module.

**Reply by zackschuster** [reply]
also iirc this is logistically unfeasible since node would have to pre-parse code to detect module syntax & thus know whether to target the Module or Script parse goal in v8. this introduces (potentially very expensive) startup delays for every individual file that node works with, regardless of whether the program even uses modules.

**Reply by ljharb** [reply]
Additionally, Module and Script parse goals are ambiguous - it's impossible to reliably determine whether any file is a Script or a Module.

**Reply by mattfysh** [reply]
I'm not saying the import statement (from the consumer/importer) should trigger the assumption

I'm saying that when the **module author** puts an **export map** in their package.json, and the condition is `import`, then it should be assumed to be ESM. This doesn't conflict with any of the comments above.

e.g. the author of the package, adds this to their package.json:

```
"exports": {
  ".": {
    "import": "./dist/pkg.esm.js" <-- condition=import
    "require": "./dist/pkg.cjs.js" <-- condition=require
  }
}
```

The author has indicated an explicit resolution for an import statement. IMO, this should default to ESM because condition=import and NOT condition=require

I know the author could also include the entrypoint file as `/dist/pkg.esm.mjs`, using the file extension to explicitly denote the file's module format. But the community have not rallied around the mjs file extension and ecosystem tooling rarely supports mjs causing all manner of latent issues, e.g. https://github.com/aelbore/esbuild-jest/issues/38

**Reply by ljharb** [reply]
This also isn't the case, because the `import` condition could point to a CJS file that is written to assume ESM is available. For example, "require" points to a file that does not use nullish coalescing, but "import" points to a file that does.

The `.mjs` file extension is the canonical way to express it, and always will be (`"type": "module"` in package.json and `.js` is an alternative, for those that prefer aesthetics over clarity). Any tool that can't handle something as basic as file extensions (modulo examples like the one you linked, where they probably just haven't added support yet) is not a tool I'd suggest relying on.

**Reply by mattfysh** [reply]
I'm sorry, but this doesn't make any sense to me - how are nullish coalescing and ESM related?

Isn't feature/syntax support tied to the engine version, and not the module type?

And if people are using export map conditions to perform feature detection, is that not an unnatural coupling?

**Reply by ljharb** [reply]
In practice, any node version that supports native ESM will support a certain level of syntax (i didn't actually check if that includes `??`, but substitute any syntax that qualifies).

It's explicitly part of the design of "exports" that the extension (and package.json "type" field), only, always, and ever, is the determining factor for the parse goal of a file. It would be wildly inconsistent and surprising to have one condition, "import", have special semantics here.

**Reply by mattfysh** [reply]
> It would be wildly inconsistent and surprising to have one condition, "import", have special semantics here.

I don't agree with this. The notion of branching specifier resolution based on a condition of "import" already carries with it a very explicit semantic.

That is, the package has been requested by an ESM module.

Therefore, shouldn't node respond in kind?

**Reply by ljharb** [reply]
"requested by an ESM module" says precisely zero about the format being imported, by design. It could be a CJS module, or a WASM module, or in the future, some other kind of module - HTML, CSS, Rust, whatever.

Only the file itself - ie, the extension and its nearest package.json - dictates the parse goal of the module, and it MUST always be identical no matter how it's consumed.

**Reply by mattfysh** [reply]
I feel like we're going around in circles here... request from an ESM module is one half of assumption guidance. The other is that the **module author** has explicitly attached a file resolution to the `import` condition, i.e. "when importing my package FROM an ESM module, I denote that file X should be used". In this very case, we have enough information - we know the author intends to respond with an ESM module.

I don't agree with the conflation of module type (ESM) with engine features (nullish coalescing).

Can you show me an example where an author might specify an export map under an `import` condition, where they would intend for the target file to NOT be an ESM module?

**Reply by ljharb** [reply]
The module author could have pointed to it via "require" in a different "exports" entry. What should happen then?

The "exports" field can't possibly have any reliable impact on how a file is parsed, and whether you agree or not, that was always an explicit part of the entire design of ESM in node.

I can only provide a contrived example, but that doesn't change that *a file must always be parsed the same, no matter how it is consumed*, which means only the file extension can determine that.

**Reply by guybedford** [reply]
> also iirc this is logistically unfeasible since node would have to pre-parse code to detect module syntax & thus know whether to target the Module or Script parse goal in v8. this introduces (potentially very expensive) startup delays for every individual file that node works with, regardless of whether the program even uses modules.

One thing that did change here since the original decisions made by the modules group is that all CommonJS modules are now loaded through the CJS lexer when being imported for export detection. It would be thus possible to extend the lexer with ESM syntax detection. There are some pipeline complexities here since it would be a late change of format to integrate into the hook model. And the ambiguous case would still apply. But it might actually now be a possible future addition without a significant perf regression, which wasn't previously the case.

**Reply by bmeck** [reply]
We have been in long arguments about parsing and the issues of doing so, I don't think we should go down that route again without good cause.

**Reply by mattfysh** [reply]
Thanks @bmeck - I do understand that it's been a contentious issue in the past. From my perspective, I don't understand the arguments against this change. It seems like such a cut and dry issue for me, as I've laid out in the reasons above. Perhaps if we don't want to use export map conditions as explicit hints to module formats, then we could try to drive more community tooling support for file extensions such as .mjs

**Reply by bmeck** [reply]
@mattfysh the general issue is as @ljharb described. import/require resolution routing are not really ways to hint at module formats. The final destination determines the format. I'd agree support for both .cjs and .mjs could certainly be better, especially as I see more .cjs files and it seems to be forgotten about quite often.

**Reply by aduh95** [reply]
@mattfysh Maybe we can try to go into the details of your proposal to see what's achievable and what are the blocking point to keep the discussion productive.
Here's what I understand your proposal is (please correct me if I got it wrong):

---

When a user tries to load `pkg/foo`:
 - If the `pkg/foo` is mapped to a `.js` file in the `package.json` `"exports"` via the `"import"` condition;
 - If the `package.json` doesn't include `"type": "module"` nor `"type": "commonjs"`;
Then the parsing goal should be implicitly ESM.

---

So for a package like the following:

```json
{
  "name": "pkg",
  "exports": { "./foo": { "import": "./file.js", "require": "./file.js" } }
}
```

`import("pck/foo")` loads `file.js` as ESM but `require("pck/foo")` or `require('/absolute/path/to/file.js')` loads it as CJS, correct?

**Reply by mattfysh** [reply]
Yes! I was with you there until the end, I didn't understand where `require('/absolute/path/to/file.js')` came from.

Also this confuses me:

```
"exports": { "./foo": { "import": "./file.js", "require": "./file.js" } }
```

Why not just

```
"exports":  {"./foo": "./file.js"}
```

I would anticipate that the point of routing by condition (import, require, etc) is a hope to route the request to a different file, not the same file.

**Reply by aduh95** [reply]
Let me try to use another example then:
```json
{
  "name": "pkg",
  "exports": { "./foo": { "import": "./file.js" }, "./bar": { "require": "./file.js" }, "./baz": "./file.js"  }
}
```
If your proposal was implemented, `import("pkg/foo")` would parse `file.js` as ESM but `require("pkg/bar")` would parse it as CJS.

Currently `require('/absolute/path/to/file.js')` or `import('file:///absolute/path/to/file.js')` would both loads `file.js` as CJS, and I'm mentioning it to make clear that your proposal don't intend to change that.

It's not clear to me what you are suggesting what should happen for `import('pkg/baz')` though...

> I would anticipate that the point of routing by condition (import, require, etc) is a hope to route the request to a different file, not the same file.

I understand that of course, I would expect if your proposal were in place that we would advice against doing that in the docs.

**Reply by mattfysh** [reply]
> If your proposal was implemented, `import("pkg/foo")` would parse `file.js` as ESM but `require("pkg/bar")` would parse it as CJS.

This is correct - but I argue that this implementation of an export map is self-contradictory, if you believe that imports and requires should ever route to the same canonical file (which by definition, can only be in one module format)

The point being (in the presence of an export map): resolving a specifier under `condition=import` should carry different semantics to resolving a specifier under `condition=require`. Specifically, when resolving under either of these conditions you should be able to safely make guaranteed assumptions about the request context, and therefore apply this same assumption to the response module type.

**Reply by ljharb** [reply]
There are potentially many specifiers that point to a file, and the one-of-many specifier that got you there isn't, shouldn't, and must not be relevant to how the file is treated.

**Reply by mattfysh** [reply]
That may be true, but the presence of an export map has the tendency of being the authoritative record. For instance, trying to resolve `import 'pkg/file'` or `require('pkg/file')` - when an export map is present - results in an exception, even if `node_modules/pkg/file.js` exists on disk.

Whereas without the export map, no error would occur.

Therefore, I argue that the export map is authoritative and the conditions expressed within can be respected to the degree that associated assumptions are fully-binding.

**Reply by ljharb** [reply]
The problem is not that the export map isn't authoritative - it's that a *single conditions object* is not authoritative.

Imagine this (assume no "type" field is present):
```json
"exports": {
  "./a": {
      "import": "./x.js"
   },
  "./b": {
      "require": "./x.js"
   }
}
```
What kind of module is `x.js`?

The answer is either "based solely on the file extension/type field", or "you can't possibly know, so it will change based on how you get to it". The latter is unacceptably inconsistent and is not authoritative at all.

**Reply by mattfysh** [reply]
This is the same example that @aduh95 posted earlier, and my response is that this type of export map is self-contradictory if you respect the explicit semantics carried by "import" and "require".

**Reply by aduh95** [reply]
> The point being: resolving a specifier under `condition=import` should carry different semantics to resolving a specifier under `condition=require`.

I understand why you may think that, but also things are so much simpler if we decide that not.
The current implementation ensures that the parsing goal of a `.js` file is always explicit. Your proposal removes that guarantee for very little benefit.

I think what you want to do is to move your exported ESM file in a subfolder folder and add a `package.json` containing `{ "type": "module" }`.

```json
// package.json
{
 "exports": {
   ".": {
     "import": "./dist/esm/pkg.js", <-- condition=import
     "require": "./dist/pkg.cjs.js" <-- condition=require
   }
 }, ...
}
```

```json
// dist/esm/package.json
{ "type": "module" }
```

If you feel strongly about that, you could set up a loader hook (still experimental) to implement your proposal.

**Reply by mattfysh** [reply]
It's a shame that `/dist/esm/package.json` is required, when module type can safely be determined by the conditions within the export map, but I will respect that this is where we disagree.

For reasons that I do not understand, there seems to be a will to accommodate that a file resolved under the export map, in an explicit and _author-specified_ `condition=import` route, can be a CJS module.

**Reply by ljharb** [reply]
@mattfysh here's another valid example, that must be valid to allow seamless composition of export map features:
```json
"exports": {
  "./a": {
      "import": {
         "require": "./x.js"
       }
    }
}
```
is this ESM or CJS?

**Reply by bmeck** [reply]
I'm just here to state conditions are no applied to the target, conditions are parameters of the loaders. Currently this is like the following in pseudocode:

```mjs
// imagine load() ~= import and require in an abstract sense
function load(specifier, dependentLocation) {
  const target = find(specifier, dependentLocation, ['node', 'require' | 'import']);
  // ...
}
```

If we instead make conditions properties of the target

```mjs
// imagine load() ~= import and require in an abstract sense
function load(specifier, dependentLocation) {
  const target = find(specifier, dependentLocation, ['node', 'require' | 'import']);
  if (target.usedConditionsToLoad.includes('import') && target.package.type == null) target.format = ...
  // ...
}
```

It causes a variety of confusions, such as:

* using resolution that doesn't go through export maps such as absolute locations (passing import.meta.url in particular causes this). in such cases `usedConditionsToLoad` would be empty always.
* to avoid that you need to know how to reverse lookup the conditions from a target (idk honestly unless we start doing full crawls of fs resolutions w/ early errors). per examples above this basically requires each target to be statically known and problems like having both an export and import pointing to the same file would have to cause early errors. This is a fair bit more work to do per load.

These manifest in ways like the ambiguity of not knowing what format a file is statically. In general, to avoid these categories of problems, the loading mechanism provides data that is parameterized but does not modify the target is the point.
